### PR TITLE
feat: allow partially failing batch inserts

### PR DESCRIPTION
**Pull request by hperl**
When batch-inserting multiple identities, conflicts or validation errors of a subset of identities in the batch still allow the rest of the identities to be inserted. The returned JSON contains the error details that lead to the failure.

## Related issue(s)

Docs PR: https://github.com/ory/docs/pull/1864

## Checklist

- [ ] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [ ] I have referenced an issue containing the design document if my change
      introduces a new feature.
- [ ] I am following the
      [contributing code guidelines](../blob/master/CONTRIBUTING.md#contributing-code).
- [ ] I have read the [security policy](../security/policy).
- [ ] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security vulnerability, I
      confirm that I got the approval (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [ ] I have added tests that prove my fix is effective or that my feature
      works.
- [ ] I have added or changed [the documentation](https://github.com/ory/docs).

### CODE REVIEW

File: identity/handler_test.go [hunk 1]
```diff
@@ -774,34 +774,49 @@ func TestHandler(t *testing.T) {
...
+					}
 
 					req := &identity.BatchPatchIdentitiesBody{Identities: patches}
-					send(t, adminTS, "PATCH", "/identities", tt.expectStatus, req)
+					body := send(t, adminTS, "PATCH", "/identities", http.StatusOK, req)
+					var actions []string
+					for _, a := range body.Get("identities.#.action").Array() {
+						actions = append(actions, a.String())
+					}
+					assert.Equal(t, []string{"create", "create", "create", "error", "create", "create"}, actions, body)
```
Comment (zepatrik) [inline identity/handler_test.go:819]: It would be good IMO to assert also the error here, just to make sure that the test does what it should do.

**Review by zepatrik — DISMISSED** [review]

File: identity/handler.go [hunk 2]
```diff
@@ -617,13 +617,22 @@ func (h *Handler) batchPatchIdentities(w http.ResponseWriter, r *http.Request, _
...
+	if err != nil && !errors.As(err, &partialErr) {
 		h.r.Writer().WriteError(w, r, err)
 		return
 	}
 	for resIdx, identitiesIdx := range indexInIdentities {
 		if identitiesIdx != nil {
-			res.Identities[resIdx].IdentityID = &identities[*identitiesIdx].ID
+			ident := identities[*identitiesIdx]
+			// Check if the identity was created successfully.
+			if failed := partialErr.Find(ident); failed != nil {
```
[the remark this answers is not in the document]
Comment (hperl) [reply identity/handler.go:630]: Fixed

File: persistence/sql/batch/create.go [hunk 3]
```diff
@@ -211,23 +262,55 @@ func Create[T any](ctx context.Context, p *TracerConnection, models []*T) (err e
 	if err != nil {
 		return sqlcon.HandleError(err)
 	}
-	defer rows.Close()
```
Comment (zepatrik) [inline persistence/sql/batch/create.go:214]: With some of the error handling we could now return early and the rows will not be closed. The documentation suggests that you are allowed to close the rows multiple times, so I'd keep the defer statement here just to ensure they are always closed.

> Close is idempotent and does not affect the result of `Rows.Err`. [from here](https://pkg.go.dev/database/sql#Rows.Close)

> Failing to read all rows or use rows.Close() reserves connections from the pool. [from here](http://go-database-sql.org/surprises.html)

File: persistence/sql/batch/create.go [hunk 4]
```diff
@@ -236,43 +319,43 @@ func Create[T any](ctx context.Context, p *TracerConnection, models []*T) (err e
...
+	if len(partialConflictError.Failed) > 0 {
+		return sqlcon.ErrUniqueViolation.WithWrap(&partialConflictError)
+	}
+
+	return nil
+
 }
 
 // setModelID was copy & pasted from pop. It basically sets
 // the primary key to the given value read from the SQL row.
```
Comment (zepatrik) [inline persistence/sql/batch/create.go:351]: I guess the implementation now does not follow pop anymore.

**Review by zepatrik — CHANGES_REQUESTED** [review]
Looks great now, just some small nits.

File: persistence/sql/batch/create.go [hunk 5]
```diff
@@ -213,21 +264,55 @@ func Create[T any](ctx context.Context, p *TracerConnection, models []*T) (err e
...
+}
+
+func handleFullInserts[T any](models []*T, rows *sql.Rows) error {
 	// Hydrate the models from the RETURNING clause.
-	//
-	// Databases not supporting RETURNING will just return 0 rows.
-	count := 0
-	for rows.Next() {
+	for i := 0; rows.Next(); i++ {
 		if err := rows.Err(); err != nil {
```
Comment (alnr) [inline persistence/sql/batch/create.go:285]: This check is incorrect AFAIK. `rows.Err()` should be checked after `rows.Next() == false`.

Comment (hperl) [reply persistence/sql/batch/create.go]: fixed :)

**Review by zepatrik — APPROVED** [review]

**Pull request merged**